### ISSUE TITLE
SuperAnimal, GPU usage, Python crashing during training

### ISSUE THREAD

**Issue by Kpraetori** [DeepLabCut/DeepLabCut#2702]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

This is separate from [Issue 2639](https://github.com/DeepLabCut/DeepLabCut/issues/2639) as it's training not analysis.
I think I sorted through the batch size bug(changed the freeze bin stats, line 48, in the FasterRCNN.py and line 110 in pytorch_config) and with installation issue (change the order of installing things), but my SuperAnimal-Quadruped model is not training using the GPU and crashes or quits out.

I checked that torch.cuda is available shows as true, and nvidia-smi shows my version. I've checked that my pytorch is an acceptable version. It's using around the same minimal amount of CPU and GPU with GPU making a few bursts for a a second before dropping again (see screenshot). 

![Screenshot 2024-08-01 at 2 34 35 PM](https://github.com/user-attachments/assets/f66b3c2b-af06-49e8-aa64-66a31c1a1960)
![Screenshot 2024-08-01 at 2 34 50 PM](https://github.com/user-attachments/assets/0cebb461-2e9e-4fa3-b33c-18ad838138e6)

I don't know what to do or change. Any help or advice is much appreciated

### Operating System

Windows 10 Enterprise

### DeepLabCut version

dlc version-3.0.0rc3

### DeepLabCut mode

single animal

### Device type

gpu, NVIDIA RTX A4000 w 128Gb RAM
NVIDIA-SMI 537.42, Driver Version: 537.42, CUDA Version: 12.2

### Steps To Reproduce

<code>
conda create -n deeplabcut3 python=3.11
conda activate deeplabcut3
conda install cuda -c nvidia/label/cuda-12.2.0
conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
conda install cudnn -c conda-forge
conda install -c conda-forge pytables==3.8.0
pip install "git+https://github.com/DeepLabCut/DeepLabCut.git@pytorch_dlc#egg=deeplabcut[gui,modelzoo,wandb]"
</code>

Go to "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\models\detectors\fasterRCNN.py"
change freeze_bin_stats to TRUE in for detector (line 48). I also went to my pytorch_config file and made sure the backbone was false, line 110 but the detector was true. And changes to image scale inputs given larger images.

### Relevant log output

```shell
(deeplabcut3) C:\>python -m deeplabcut
Loading DLC 3.0.0rc3...
Starting GUI...
QPixmap::scaled: Pixmap is a null pixmap
Training with configuration:
data:
  colormode: RGB
  inference:
    normalize_images: True
    auto_padding:
      pad_width_divisor: 32
      pad_height_divisor: 32
  train:
    affine:
      p: 0.5
      rotation: 30
      scaling: [1.0, 1.0]
      translation: 0
    collate: None
    covering: False
    gaussian_noise: 12.75
    hist_eq: False
    motion_blur: True
    normalize_images: True
detector:
  data:
    colormode: RGB
    inference:
      normalize_images: True
    train:
      affine:
        p: 0.5
        rotation: 30
        scaling: [1.0, 1.0]
        translation: 40
      collate:
        type: ResizeFromDataSizeCollate
        min_scale: 0.2
        max_scale: 7.0
        min_short_side: 128
        max_short_side: 1152
        multiple_of: 32
        to_square: False
      hflip: True
      normalize_images: True
  device: auto
  model:
    type: FasterRCNN
    freeze_bn_stats: True
    freeze_bn_weights: False
    variant: fasterrcnn_resnet50_fpn_v2
  runner:
    type: DetectorTrainingRunner
    eval_interval: 1
    optimizer:
      type: AdamW
      params:
        lr: 0.0001
    scheduler:
      type: LRListScheduler
      params:
        milestones: [160]
        lr_list: [[1e-05]]
    snapshots:
      max_snapshots: 5
      save_epochs: 40
      save_optimizer_state: False
  train_settings:
    batch_size: 8
    dataloader_workers: 0
    dataloader_pin_memory: True
    display_iters: 1000
    epochs: 200
device: auto
metadata:
  project_path: E:\KelseyW\DogPose Copies\July30\DogPose-KW-2024-04-03
  pose_config_path: E:\KelseyW\DogPose Copies\July30\DogPose-KW-2024-04-03\dlc-models-pytorch\iteration-0\DogPoseApr3-trainset80shuffle1\train\pose_cfg.yaml
  bodyparts: ['nose', 'left_eye', 'right_eye', 'front_left_paw', 'front_right_paw', 'back_left_paw', 'back_right_paw', 'left_earbase', 'right_earbase', 'left_earend', 'right_earend', 'neck_base', 'neck_end', 'throat_base', 'throat_end', 'left_hip', 'right_hip', 'back_end', 'tail_taper', 'tail_end']
  unique_bodyparts: []
  individuals: ['animal']
  with_identity: None
method: td
model:
  backbone:
    type: HRNet
    model_name: hrnet_w32
    freeze_bn_stats: False
    freeze_bn_weights: False
    interpolate_branches: False
    increased_channel_count: False
  backbone_output_channels: 32
  heads:
    bodypart:
      type: HeatmapHead
      weight_init: normal
      predictor:
        type: HeatmapPredictor
        apply_sigmoid: False
        clip_scores: True
        location_refinement: False
      target_generator:
        type: HeatmapGaussianGenerator
        num_heatmaps: 20
        pos_dist_thresh: 17
        heatmap_mode: KEYPOINT
        generate_locref: False
      criterion:
        heatmap:
          type: WeightedMSECriterion
          weight: 1.0
      heatmap_config:
        channels: [32]
        kernel_size: []
        strides: []
        final_conv:
          out_channels: 20
          kernel_size: 1
net_type: hrnet_w32
runner:
  type: PoseTrainingRunner
  gpus: None
  key_metric: test.mAP
  key_metric_asc: True
  eval_interval: 10
  optimizer:
    type: AdamW
    params:
      lr: 0.0001
  scheduler:
    type: LRListScheduler
    params:
      lr_list: [[1e-05], [1e-06]]
      milestones: [160, 190]
  snapshots:
    max_snapshots: 5
    save_epochs: 40
    save_optimizer_state: False
train_settings:
  batch_size: 8
  dataloader_workers: 0
  dataloader_pin_memory: True
  display_iters: 1000
  epochs: 200
  seed: 42
  weight_init:
    dataset: superanimal_quadruped
    with_decoder: False
    memory_replay: False
Loading detector checkpoint from C:\Other_Program_Files\miniforge3\envs\deeplabcut3\Lib\site-packages\deeplabcut\modelzoo\checkpoints\superanimal_quadruped_fasterrcnn_parsed.pt
C:\Other_Program_Files\miniforge3\envs\deeplabcut3\Lib\site-packages\deeplabcut\pose_estimation_pytorch\data\transforms.py:68: UserWarning: Be careful! Do not train pose models with horizontal flips if you have symmetric keypoints!
  warnings.warn(
Data Transforms:
  Training:   Compose([
  HorizontalFlip(always_apply=False, p=0.5),
  Affine(always_apply=False, p=0.5, interpolation=1, mask_interpolation=0, cval=0, mode=0, scale={'x': (1.0, 1.0), 'y': (1.0, 1.0)}, translate_percent=None, translate_px={'x': (0, 40), 'y': (0, 40)}, rotate=(-30, 30), fit_output=False, shear={'x': (0.0, 0.0), 'y': (0.0, 0.0)}, cval_mask=0, keep_ratio=True, rotate_method='largest_box'),
  Normalize(always_apply=False, p=1.0, mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], max_pixel_value=255.0),
], p=1.0, bbox_params={'format': 'coco', 'label_fields': ['bbox_labels'], 'min_area': 0.0, 'min_visibility': 0.0, 'min_width': 0.0, 'min_height': 0.0, 'check_each_transform': True}, keypoint_params={'format': 'xy', 'label_fields': ['class_labels'], 'remove_invisible': False, 'angle_in_degrees': True, 'check_each_transform': True}, additional_targets={}, is_check_shapes=True)
  Validation: Compose([
  Normalize(always_apply=False, p=1.0, mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], max_pixel_value=255.0),
], p=1.0, bbox_params={'format': 'coco', 'label_fields': ['bbox_labels'], 'min_area': 0.0, 'min_visibility': 0.0, 'min_width': 0.0, 'min_height': 0.0, 'check_each_transform': True}, keypoint_params={'format': 'xy', 'label_fields': ['class_labels'], 'remove_invisible': False, 'angle_in_degrees': True, 'check_each_transform': True}, additional_targets={}, is_check_shapes=True)
Using custom collate function: {'type': 'ResizeFromDataSizeCollate', 'min_scale': 0.2, 'max_scale': 7.0, 'min_short_side': 128, 'max_short_side': 1152, 'multiple_of': 32, 'to_square': False}
Using 628 images and 157 for testing

Starting object detector training...
--------------------------------------------------
```

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by n-poulsen** [comment]
Hi @Kpraetori! I'm not sure I understood - does the code crash while training, or is it just slow to use?

**Comment by Kpraetori** [comment]
Hi @n-poulsen thanks for getting back to me. Both? It acts like it's running very slowly, as in the screenshots, but it's crashed and the GUI says not responding in the title bar. If I click on the GUI, the mouse wheel spins, then the not responding announcement will come up and the app will close.

**Comment by n-poulsen** [comment]
Is anything being printed in the terminal?

**Comment by Kpraetori** [comment]
@n-poulsen It is the standard starting training message but in the top of the window it says not responding.

I let it sit for a few hours once and it was still frozen and nothing changed. It wasn't a matter of being slow.

**Comment by n-poulsen** [comment]
I think the issue might be that the GPU is not being used, and the load is too high for the CPU so it just freezes. First, could you open an `ipython` shell in your Anaconda prompt and check if PyTorch can use the GPU? You can do so by running:

```python
import torch
print(torch.cuda.is_available())
```

If that prints True, could you try with display iterations set to 1 (no need to run this for long - if anything is happening there should be lines printed quickly, in less than a minute). You could also reduce the computational load and see if something is happening by updating your model configuration file (the `pytorch_config.yaml`) and setting the collate parameters to:

```
...
detector:
  ...
  data:
    ...
    train:
      ...
      collate:
        type: ResizeFromDataSizeCollate
        min_scale: 0.1
        max_scale: 0.5
        min_short_side: 128
        max_short_side: 256
        multiple_of: 32
        to_square: False
      ...
```

This will make the images given to the model smaller, which means it should be much faster to run them (quick note: **do NOT do this to train your model - here I just want to check if everything freezes up even with a lower computational load**). In your detector's `train_settings`, you could also set `batch_size: 2` to further reduce the load.

**Comment by Kpraetori** [comment]
Okay, thanks for helping me troubleshoot!

```
print(torch.cuda.is_available())
True
```
So that's working.
I made all the changes you suggested and it has started running at least, but seems to be using the CPU rather/more than GPU. 
<img width="1435" alt="Screenshot 2024-09-03 at 10 47 28" src="https://github.com/user-attachments/assets/1fe3317e-9b70-4617-97d2-c30b4807de58">

<img width="1435" alt="Screenshot 2024-09-03 at 10 48 53" src="https://github.com/user-attachments/assets/93033624-1237-4b6a-9c6a-72ac2b982802">

Any idea how to get it to use the GPU it has access to over the CPU?

**Comment by Kpraetori** [comment]
@n-poulsen I rebuilt it just in case something went awry with that last install. 

EDIT WITH UPDATE: With this setup it is (slowly) running the model. I've left it be for the last 2 days and just now it started the second model. Yay! 

I did install in a slightly different order so maybe that played a role? Is there something I should change for refinining the model?

My Steps:

```
conda create -n deeplabcut3 python=3.11
conda activate deeplabcut3
conda install cuda -c nvidia/label/cuda-12.2.0
conda install cudnn -c conda-forge
conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
conda install -c conda-forge pytables==3.8.0
pip install "git+https://github.com/DeepLabCut/DeepLabCut.git@pytorch_dlc#egg=deeplabcut[gui,modelzoo,wandb]"
```

My checks:
```
>nvcc -V
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Tue_Jun_13_19:42:34_Pacific_Daylight_Time_2023
Cuda compilation tools, release 12.2, V12.2.91
Build cuda_12.2.r12.2/compiler.32965470_0
```
And
```
Python 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:27:10) [MSC v.1938 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> print(torch.cuda.is_available())
True
```
I tried with the settings I mentioned before, and this is what it looks like:
<img width="1435" alt="Screenshot 2024-09-03 at 14 40 10" src="https://github.com/user-attachments/assets/47732404-bdf6-42fd-b292-d07feb3489eb">

No new errors or anything. I got a single epoch when I checked an hour later, so it's still not using GPU properly. I've restarted with batch size of 4 and will see how far it gets. [Edited for grammar.]

**Comment by n-poulsen** [comment]
Great news that you managed to install without errors! I believe your GPU is being used (dedicated GPU memory is 15.8GB out of 16GB). Are you tracking metrics over time (e.g. with `wandb`)?

I see that you're training for 200 epochs and that might be a lot in this case, as you're just fine-tuning a model instead of training from scratch (it **shouldn't** hurt performance, but you might be able to save a lot of time by not training as long). Keeping an eye on losses/metrics is a way to check that your model is still learning, and that it's useful to continue training.

The current SuperAnimal models we have are powerful but not very fast, we're working on faster models which should be available very soon, so that might be a reason why training is slow. One thing that might bring a speedup is increasing the number of DataLoader workers. You can do that by editing your `pytorch_config.yaml` so that it contains:

```yaml
detector:
  ...
  train_settings:
    ...
    batch_size: 8
    dataloader_workers: 2
    dataloader_pin_memory: False
...
train_settings:
  ...
  batch_size: 16
  dataloader_workers: 4
  dataloader_pin_memory: False
```
